### PR TITLE
checkpoint: resolve symlink for external bind mount

### PR DESCRIPTION
**Pull request by liusdu**
runc resolves symlink before doing bind mount. So
we should save original path while formatting CriuReq for
checkpoint.

Signed-off-by: Liu Hua <weldonliu@tencent.com>

### CODE REVIEW

**Comment by kolyshkin** [comment]
@liusdu thank you for the patch! Nice that you're using checkpoint/restore!

Can you add a test case (to e.g. `tests/integration/checkpoint.bats`?)

**Comment by kolyshkin** [comment]
@liusdu ^^^

**Comment by liusdu** [comment]
> @liusdu ^^^

@kolyshkin  Sorry for this late,  checkpoint/restore  of runc works very well for me.

File: tests/integration/checkpoint.bats
Comment (kolyshkin) [inline tests/integration/checkpoint.bats:132]: Not: missing space before `(`

Comment (liusdu) [reply tests/integration/checkpoint.bats]: Thans.. Fixed..

**Comment by liusdu** [comment]
@kolyshkin

File: tests/integration/checkpoint.bats
Comment (kolyshkin) [inline tests/integration/checkpoint.bats:141]: nit: extra empty line

File: libcontainer/container_linux.go
Comment (kolyshkin) [inline libcontainer/container_linux.go:782]: nit: extra empty line

**Comment by kolyshkin** [comment]
LGTM except for a couple of nits. @adrianreber PTAL

**Comment by adrianreber** [comment]
@liusdu Can you give an example where you saw this error. It looks correct, but I am curious when this can happen.

**Comment by liusdu** [comment]
@adrianreber  Look at following Dockerfile

```
FROM openjdk:8-jre
RUN wget  -qO-  'https://www.apache.org/dyn/closer.cgi?action=download&filename=flink/flink-1.12.2/flink-1.12.2-bin-scala_2.11.tgz' |tar -zvx -C /opt
RUN  ln -s  /opt/flink-1.12.2  /opt/flink
```

If we pass configuration as volume to container, we can meet this type of error.

`docker run -d -v  ${FLINK_CONF}`:/opt/flink/conf. ${flink_image}`

**Comment by liusdu** [comment]
@kolyshkin thanks for you review~

File: tests/integration/checkpoint.bats
Comment (kolyshkin) [inline tests/integration/checkpoint.bats:132]: please add an empty line before @test

Comment (liusdu) [reply tests/integration/checkpoint.bats:133]: Fixed

**Comment by liusdu** [comment]
ping @kolyshkin @adrianreber

**Review by kolyshkin — APPROVED** [review]
LGTM

**Comment by kolyshkin** [comment]
close/reopen to re-kick CI

File: tests/integration/checkpoint.bats
Comment (haircommander) [inline tests/integration/checkpoint.bats:135]: This is not blocking, but just curious:
why an absolute path for `/real/conf`? I see we're using the same path in `mounts.bats`, so it works, but it's not totally clear to me why we're using this path rather than somewhere in a tmpfs (or at least something test-specific)

Comment (liusdu) [reply tests/integration/checkpoint.bats:135]: Hi @haircommander 
Commit [runc run: resolve tmpfs mount dest in container scope](https://github.com/opencontainers/runc/commit/637f82d6b8b45a99519b863f8cf7461760da7a97)  changes behavior of tmpfs mount and bind mount, when the destination is symlink.
 -  changes of bind mount breaks the functionality of runc checkpoint. This is why we need this path.
 -  changes of tmpfs mount does not break anything. Since tmpfs mount is fully owned by container's mnt namespace.

Comment (haircommander) [reply tests/integration/checkpoint.bats:135]: thanks for the explanation, makes sense

File: libcontainer/container_linux.go
Comment (cyphar) [inline libcontainer/container_linux.go:783]: I do wonder -- what happens if the mount is moved from inside the container (with a simple `rename`) after the filesystem has been set up? Is this also something CRIU doesn't support / really doesn't like?

Comment (liusdu) [reply libcontainer/container_linux.go:783]: Yes @cyphar  Since CRIU does not know how to restore bind mounts(master in host namespace),  CRIU need  this type of mount marked as external(via `--external`). 
  runc formats `--external` parameter according to spec((link)[https://github.com/opencontainers/runc/blob/master/libcontainer/container_linux.go#L1112]). So for situation you metioned,  criu does not work properly.  btw, should we export `--external` for runc. Then users can mark external resource themselves?

Comment (adrianreber) [reply libcontainer/container_linux.go:783]: >> I do wonder -- what happens if the mount is moved from inside the container (with a simple rename) after the filesystem has been set up? Is this also something CRIU doesn't support / really doesn't like?

I do not understand this question. But initially I would have said that CRIU can handle it. Now @liusdu says it is not possible. I probably misunderstood the question.

> So for situation you mentioned, CRIU does not work properly. btw, should we export `--external` for runc. Then users can mark external resource themselves?

For Kubernetes, restoring containers in different Pods, I will need CRIU's `--external` to be available at the runc level.

Currently runc informs CRIU about all external mounts but it is not possible to change the location of the external mount between checkpoint and restore. For Kubernetes I will need to change the location of an external mount. The most obvious bind mount currently is the location of the secrets directory which has the Pod UID in its container external path and I need to tell CRIU/runc that that directory should be mounted from some other location.

Comment (liusdu) [reply libcontainer/container_linux.go:783]: ping @cyphar  ..

**Comment by liusdu** [comment]
@haircommander can you approve this pr?

**Comment by haircommander** [comment]
It LGTM, but I'm not a maintainer so my approval wouldn't help unfortunately
@kolyshkin @mrunalp @AkihiroSuda PTAL

**Review by kolyshkin — APPROVED** [review]
LGTM

**Comment by kolyshkin** [comment]
We're in the middle of 1.0.0 GA release. Once it's done, I think we can merge this.

**Review by mrunalp — APPROVED** [review]
LGTM

**Comment by kolyshkin** [comment]
There's something wrong with CI :(

**Comment by kolyshkin** [comment]
Might be caused by https://github.com/opencontainers/runc/pull/3029. @liusdu can you take a look? The failed CI can be seen from the Actions tab, https://github.com/opencontainers/runc/actions

**Pull request merged**
